src/crimson/os/seastore/seastore_types.h

```diff
@@ -36,28 +36,148 @@ struct seastore_meta_t {
   }
 };
 
+// internal segment id type of segment_id_t, basically
+// this is a unsigned int with the top "device_id_len"
+// bits representing the id of the device on which the
+// segment resides
+using internal_segment_id_t = uint32_t;
+// type of device id
+using device_id_t = uint8_t;
+// order of device_id_t
+constexpr uint16_t device_id_len = 4;
+// maximum devices supported
+constexpr uint16_t max_devices = 1 << device_id_len;
+
+// segment ids without a device id encapsulated
+using device_segment_id_t = internal_segment_id_t;
+
+// mask for segment manager id
+constexpr internal_segment_id_t SM_ID_MASK =
+  0xF << (std::numeric_limits<internal_segment_id_t>::digits - device_id_len);
+// default internal segment id
+constexpr internal_segment_id_t DEFAULT_INTERNAL_SEG_ID =
+  std::numeric_limits<internal_segment_id_t>::max() - 1;
+
 // Identifies segment location on disk, see SegmentManager,
-using segment_id_t = uint32_t;
+struct segment_id_t {
+  internal_segment_id_t segment = DEFAULT_INTERNAL_SEG_ID;
+
+  segment_id_t() = default;
+  constexpr segment_id_t(internal_segment_id_t segment)
+    : segment(segment) {}
+  segment_id_t(device_id_t id, internal_segment_id_t segment)
+    : segment(add_device_id(segment, id)) {
+    // only lower 4 bits are effective, and we have to reserve 0x0F for
+    // special XXX_SEG_IDs
+    assert(id < 15);
+  }
+
+  [[gnu::always_inline]]
+  device_id_t device_id() const {
+    return get_device_id(segment);
+  }
+
+  [[gnu::always_inline]]
+  internal_segment_id_t internal_segment_id() const {
```

Comment:
should use device_segment_id_t to be clearly understand even though they are same type.

Reply:
Correct, literally nothing outside of this class should see internal_segment_id_t.

Reply:
Rename this to get_device_segment_id() and have it return device_segment_id_t as suggested by @liu-chunmei elsewhere.

Reply:
will do